datastore/kubernetes.go

```diff
@@ -243,11 +245,69 @@ func (s *DataStore) UpdatePod(obj *corev1.Pod) (*corev1.Pod, error) {
 	return s.kubeClient.CoreV1().Pods(s.namespace).Update(context.TODO(), obj, metav1.UpdateOptions{})
 }
 
+// CreateLease creates a Lease resource for the given CreateLease object
+func (s *DataStore) CreateLease(lease *coordinationv1.Lease) (*coordinationv1.Lease, error) {
+	return s.kubeClient.CoordinationV1().Leases(s.namespace).Create(context.TODO(), lease, metav1.CreateOptions{})
+}
+
+// GetLease gets the Lease for the given name
+func (s *DataStore) GetLeaseRO(name string) (*coordinationv1.Lease, error) {
+	return s.leaseLister.Leases(s.namespace).Get(name)
+	// return s.kubeClient.CoordinationV1().Leases(namespace).Get(context.Background(), name, metav1.GetOptions{})
+}
+
+// GetLease returns a new Lease object for the given name
+func (s *DataStore) GetLease(name string) (*coordinationv1.Lease, error) {
+	resultRO, err := s.GetLeaseRO(name)
+	if err != nil {
+		return nil, err
+	}
+	// Cannot use cached object from lister
+	return resultRO.DeepCopy(), nil
+}
+
+// ListLeasesRO returns a list of all Leases for the given namespace,
+// the list contains direct references to the internal cache objects and should not be mutated.
+func (s *DataStore) ListLeasesRO() ([]*coordinationv1.Lease, error) {
+	return s.leaseLister.Leases(s.namespace).List(labels.Everything())
+}
+
 // DeleteLease deletes Lease with the given name in s.namespace
 func (s *DataStore) DeleteLease(name string) error {
 	return s.kubeClient.CoordinationV1().Leases(s.namespace).Delete(context.TODO(), name, metav1.DeleteOptions{})
 }
 
+// UpdateLease updates the Lease resource with the given object and namespace
+func (s *DataStore) UpdateLease(lease *coordinationv1.Lease) (*coordinationv1.Lease, error) {
+	return s.kubeClient.CoordinationV1().Leases(s.namespace).Update(context.TODO(), lease, metav1.UpdateOptions{})
+}
+
+// IsRWXVolumeInFailover checks whether the volume has a lease by the same name, which an RWX volume should,
+// and whether that lease's spec shows that its holder is delinquent (its acquire time has been zeroed.)
+// If so, return the delinquent holder.
+// Any hiccup yields a return of "false".
+func (s *DataStore) IsRWXVolumeInFailover(name string) (failover bool, holder string, err error) {
+	if enabled, _ := s.GetSettingAsBool(types.SettingNameEnableShareManagerFastFailover); !enabled {
+		// failover is false, holder is empty, err is nil
+		return
+	}
+
+	lease, err := s.GetLeaseRO(name)
+	if err != nil {
+		logrus.WithError(err).Warnf("Failed to get lease for RWX volume %v", name)
+		return
+	}
+
+	holder = *lease.Spec.HolderIdentity
+	if holder == "" {
+		return
+	}
+	if (lease.Spec.AcquireTime).IsZero() {
+		failover = true
```

Comment:
It is set to zero by share manager controller in markShareManagerDelinquent.

This may be obvious, but can you remind me again why we need to actively mark delinquency by zeroing the AcquireTime @james-munson? All callers to IsNodeDelinquent -> IsRWXVolumeInFailover can access the lease, so they could all just do the simple math to determine that it hasn't been renewed recently enough, right?